geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/GatewaySenderManageFunction.java

```diff
@@ -0,0 +1,148 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.management.internal.configuration.functions;
+
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.annotations.Immutable;
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.execute.FunctionContext;
+import org.apache.geode.cache.wan.GatewaySender;
+import org.apache.geode.internal.cache.execute.InternalFunction;
+import org.apache.geode.internal.cache.xmlcache.CacheXml;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.management.internal.configuration.domain.XmlEntity;
+import org.apache.geode.management.internal.functions.CliFunctionResult;
+
+public class GatewaySenderManageFunction implements InternalFunction {
+  private static final Logger logger = LogService.getLogger();
+  private static final long serialVersionUID = -6022504778575202026L;
+
+  @Immutable
+  public static final GatewaySenderManageFunction INSTANCE = new GatewaySenderManageFunction();
+
+  private static final String ID = GatewaySenderManageFunction.class.getName();
+
+  @Override
+  public boolean hasResult() {
+    return true;
+  }
+
+
+  /**
+   * This function is used to report all the gateway sender instances for a given sender
+   * if all intances are stopped. In order to do so, it invokes the setMustQueueDroppedEvents()
+   * method
+   * on every gateway sender instance for the gateway sender passed as a parameter.
+   * The function requires an Object[] as context.getArguments() in which
+   * - the first parameter must be a Boolean stating of all gateway sender instances are stopped
+   * - the second parameter must be a String containing the name of the sender.
+   */
+  @Override
+  public void execute(FunctionContext context) {
+    String memberName = context.getMemberName();
+    try {
+      Object[] arguments;
+
+      Boolean stop = getStopFromArguments(context.getArguments());
+      if (stop == null) {
+        context.getResultSender().lastResult(new CliFunctionResult("", false,
+            "Wrong arguments passed"));
+        return;
+      }
+      String senderId = getSenderIdFromArguments(context.getArguments());
+      if (senderId == null) {
+        context.getResultSender().lastResult(new CliFunctionResult("", false,
+            "Wrong arguments passed"));
+        return;
+      }
+
+      Cache cache = context.getCache();
+      GatewaySender sender = cache.getGatewaySender(senderId);
+
+      if (sender == null) {
+        context.getResultSender().lastResult(new CliFunctionResult(memberName, false,
+            String.format("Sender '%s' does not exist", senderId)));
+        return;
+      }
+
+      sender.setMustQueueDroppedEvents(!stop);
+      XmlEntity xmlEntity = new XmlEntity(CacheXml.GATEWAY_SENDER, "name", senderId);
```

Comment:
I don't think you need to modify the cache xml for this flag, there is no need to send this back to the command. All this function does is to turn on/off this flag and report the sucess/failure back to the command.

Reply:
a good example to use is `CreateDiskStoreFunction`, that one extends CliFunction which handles the exceptions for you.

Reply:
I did not want to extend CliFunction because I intended this function to be used by users of the Java API as @gesterzhou was after.

In a new commit:

- I have simplified the function a bit because the first level try/catch was already taking care of any exception.

- I have also changed the name of the setter/flag from setMustQueueDroppedEvents to setAllInstancesStopped. I think it is a more understandable name without knowing the low level details about the dropped events.

@gesterzhou Would this solution be acceptable for you having fixed the problem in gfsh and at the same time having a function that could be called by Java API users?

Reply:
@jinmeiliao Any idea why org.apache.geode.internal.cache.wan.wancommand.StopGatewaySenderCommandDUnitTest > testStopGatewaySender_Group tests are failing?

Reply:
> @jinmeiliao Any idea why org.apache.geode.internal.cache.wan.wancommand.StopGatewaySenderCommandDUnitTest > testStopGatewaySender_Group tests are failing?

The function you wrote `GatewaySenderManageFunction` needs to be in the sanctioned-serializable list, because it is not, Geode rejected it when it's been serialized/deserialized across the wire (when you try to call it from the locator on all the servers). That's why "org.apache.geode.codeAnalysis.AnalyzeCoreSerializablesJUnitTest > testSerializables " is also failing with a new serializable class detected but not in the list. You should add that. Then both these test failures would go away.

Reply:
@jinmeiliao Thanks a lot for your support!